internal/graph/check_test.go

```diff
@@ -1940,6 +1942,85 @@ func TestUnionCheckFuncReducer(t *testing.T) {
 	})
 }
 
+func TestCheckWithFastPathOptimization(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+	usersetBatchSize := uint32(10)
+	ds := memory.New()
+	t.Cleanup(ds.Close)
+	storeID := ulid.Make().String()
+	model := testutils.MustTransformDSLToProtoWithID(`
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define viewer: [user]
+			type doc
+				relations
+					define viewer: viewer from parent
+					define parent: [folder]`)
+
+	// add some folders as parents of the document
+	maxFolderID := int(usersetBatchSize * 10)
+	for i := 0; i <= maxFolderID; i++ {
+		err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+			tuple.NewTupleKey("doc:1", "parent", fmt.Sprintf("folder:%d", i)),
+		})
+		require.NoError(t, err)
+	}
+
+	// the user is viewer of the last folder so we force early flushing of the batches when computing membership
```

Comment:
What about the case when the user is viewer of the first one? That should prevent reading the whole iterator and short circuit properly. Forcing it to go through here: https://github.com/openfga/openfga/pull/1804/files#diff-5074b7eeccca5be19ff0fa2a8f712335a6ed075628b66a46697ca9d8ca99fbe0R993

Reply:
In my test, the producer always finishes before the consumer finds an answer 🤔